rfcs/0001-adding-package-logic-to-loader.md

```diff
@@ -0,0 +1,37 @@
+- Feature Name: adding_python_package_logic_to_loader
+- Start Date: 2018-06-29
+- RFC PR:
+- Salt Issue:
+
+# Summary
+[summary]: #summary
+
+Add support for python packages to the loader. Instead of just having
+apache.py, we would have `salt/modules/apache/__init__.py`, and all the other
+modules under that directory loaded in the `apache.<function>` namespace that
+we already have.
+
+# Motivation
+[motivation]: #motivation
+
+Split modules out into different git repositories, and allow the community more control to help maintain them.
+
+# Design
+[design]: #detailed-design
+
+This should be pretty straight forward, we should be able to use the same
+_module_dirs that we are using now, and iterate over the files, and if they are
+a directory, then we load all of the modules underneath it into the namespace
+of the directory.
+
+Heirarchy would be alphabetical, and we would need to log messages if the
+function exists in an earlier file that has been loaded.
+
+## Alternatives
+[alternatives]: #alternatives
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
```

Comment:
This seems give a major impact elsewhere:

- How `setup.py` should behave from now on?
- Since this is totally different Git repos, it impacts releases, tests and maintenance. How release process will look like here, sine we will get more than one source tarball?
- If we decide to keep a single tarball on the release, how do we make sure that the modules are properly tested and aren't broken?
- Splitting repos will also split packages and this should be taken to the account.

Reply:
So, this rfc is not about splitting the modules into new repositories yet.  That will be a different one, this is just about supporting the ability to load packages/submodules into the same namespace.

The rest of these questions will be answered in a different rfc.

Reply:
Well, [you wrote it is about splitting](https://github.com/saltstack/salt/pull/48631/files#diff-983ac08a468477835dc9d9214f7d7c66R17) and that confused me. :wink: 

I would say "motivation" here is to have a namespace. And that makes totally lots of sense here (and I am all for it!).